x-pack/plugins/infra/server/lib/alerting/inventory_metric_threshold/inventory_metric_threshold_executor.ts

```diff
@@ -40,145 +41,182 @@ interface InventoryMetricThresholdParams {
-       * Uncomment the code below once they've been implemented
-       * Reference: https://github.com/elastic/kibana/issues/87048
-       */
-      // } else if (nextState === AlertStates.OK && prevState?.alertState === AlertStates.ALERT) {
-      // reason = results
-      //   .map((result) => buildReasonWithVerboseMetricName(result[item], buildRecoveredAlertReason))
-      //   .join('\n');
-    }
-    if (alertOnNoData) {
-      if (nextState === AlertStates.NO_DATA) {
-        reason = results
-          .filter((result) => result[item].isNoData)
-          .map((result) => buildReasonWithVerboseMetricName(result[item], buildNoDataAlertReason))
-          .join('\n');
-      } else if (nextState === AlertStates.ERROR) {
+
+    const logQueryFields = await libs
+      .getLogQueryFields(
+        sourceId || 'default',
+        services.savedObjectsClient,
+        services.scopedClusterClient.asCurrentUser
+      )
+      .catch(() => undefined);
+
+    const compositeSize = libs.configuration.inventory.compositeSize;
+
+    const results = await Promise.all(
+      criteria.map((condition) =>
+        evaluateCondition({
+          condition,
+          nodeType,
+          source,
+          logQueryFields,
+          esClient: services.scopedClusterClient.asCurrentUser,
+          compositeSize,
+          filterQuery,
+        })
+      )
+    );
+
+    const inventoryItems = Object.keys(first(results)!);
+    for (const item of inventoryItems) {
+      // AND logic; all criteria must be across the threshold
+      const shouldAlertFire = results.every((result) =>
+        // Grab the result of the most recent bucket
+        last(result[item].shouldFire)
+      );
+      const shouldAlertWarn = results.every((result) => last(result[item].shouldWarn));
+
+      // AND logic; because we need to evaluate all criteria, if one of them reports no data then the
+      // whole alert is in a No Data/Error state
+      const isNoData = results.some((result) => last(result[item].isNoData));
+      const isError = results.some((result) => result[item].isError);
+
+      const nextState = isError
+        ? AlertStates.ERROR
+        : isNoData
+        ? AlertStates.NO_DATA
+        : shouldAlertFire
+        ? AlertStates.ALERT
+        : shouldAlertWarn
+        ? AlertStates.WARNING
+        : AlertStates.OK;
+      let reason;
+      let threshold;
+      let value;
+      if (nextState === AlertStates.ALERT || nextState === AlertStates.WARNING) {
         reason = results
-          .filter((result) => result[item].isError)
-          .map((result) => buildReasonWithVerboseMetricName(result[item], buildErrorAlertReason))
+          .map((result) => {
+            const resultItem = result[item];
+            value = resultItem.currentValue;
+            threshold =
+              nextState === AlertStates.WARNING
+                ? resultItem.warningThreshold!
+                : resultItem.threshold;
```

Comment:
@weltenwort Can we discuss about that? I also had my doubts and indeed the values will be overwritten during the iteration

Reply:
I changed this, but what I still don't like is that I use `first(results)![item]` to grab the inventory item I am interested in. Any thoughts?

Reply:
As discussed we don't need to send threshold and value, since we have multiple conditions and it doesn't make sense to combine these values into one. The indexed reason field is going to contain the combined information from multiple conditions.